README


This archive contains the distribution Dancer-Plugin-FlashMessage,
version 0.313:

  Dancer plugin to display temporary messages, so called "flash messages".

This software is copyright (c) 2011 by Damien "dams" Krotkine.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.313     2011-02-07 10:11:55 Europe/Paris
        . fixed operators priorities

0.312     2011-02-06 13:09:54 Europe/Paris
        . depends on Dancer 1.3001

0.311     2011-02-05 21:27:23 Europe/Paris
        . fix versioning issue
        . check there is a session engine
        . update documentation

0.310     2011-02-05 18:50:27 Europe/Paris
        . fix case where no flash has been defined
        . thanks to ambs and sawyerX

0.309     2011-02-05 13:24:13 Europe/Paris
        . documentation fixed

0.308     2011-02-05 13:14:07 Europe/Paris
	. remove the persistence config
	. get the module to be as simple as possible
        . documentation fixed

0.307     2011-01-30 13:15:30 Europe/Paris
	. add Dancer dependancy
	. get Makefile.PL properly generated

0.306     2011-01-28 16:15:49 Europe/Paris
	. Fix RT#65009 again
	. flash messages are now smart enough to stay alive across multiple
	  requests, until they are used in templates ( thanks to Flavio Poletti
	  for the idea )

0.305     2011-01-23 19:20:17 Europe/Paris
	. Fix RT#65009
	. 'persistence' configuration key added
	. tests for (non) persistence added

0.304     2011-01-16 14:25:58 Europe/Paris
	. Fix tests

0.303     2011-01-15 13:58:20 Europe/Paris
	. Don't require Dancer::Plugin::TemplateToolkit (hence Template::Toolkit)
	  for tests
	. Add warning about Dist::Zilla in Makefile.PL

0.302     2011-01-11 17:41:49 Europe/Paris

0.301     2011-01-11 17:40:16 Europe/Paris

0.300     2011-01-11 17:39:00 Europe/Paris
	. Dist-Zillification

0.2
	. more Rails-like : flash is a hash of messages

0.1
	. first version


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'Dancer::Test' => '0',
    'English' => '0',
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'ExtUtils::MakeMaker' => '6.31',
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Dancer plugin to display temporary messages, so called "flash messages".',
  'dist_author' => [
    'Damien "dams" Krotkine'
  ],
  'dist_name' => 'Dancer-Plugin-FlashMessage',
  'dist_version' => '0.313',
  'license' => 'perl',
  'module_name' => 'Dancer::Plugin::FlashMessage',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Carp' => '0',
    'Dancer' => '1.3001',
    'Dancer::Plugin' => '0'
  },
  'script_files' => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer/Plugin/FlashMessage.pm
#
# This file is part of Dancer-Plugin-FlashMessage
#
# This software is copyright (c) 2011 by Damien "dams" Krotkine.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict;
use warnings;

package Dancer::Plugin::FlashMessage;
BEGIN {
  $Dancer::Plugin::FlashMessage::VERSION = '0.313';
}
# ABSTRACT: Dancer plugin to display temporary messages, so called "flash messages".

use Carp;
use Dancer ':syntax';
use Dancer::Plugin;

our $AUTHORITY = 'DAMS';

my $conf = plugin_setting;
my $token_name       = $conf->{token_name}       || 'flash';
my $session_hash_key = $conf->{session_hash_key} || '_flash';

my $session_engine;

register flash => sub ($;$) {
    my ($key, $value) = @_;

    $session_engine ||= engine 'session'
      or croak __PACKAGE__ . " error2 : there is no session engine configured in the configuration. You need a session engine to be able to use this plugin";

    my $flash = session($session_hash_key) || {};
    @_ == 2 and $flash->{$key} = $value;
    @_ == 1 and $value = delete $flash->{$key};
    session $session_hash_key, $flash;
    return $value;
};

before_template sub {
    shift->{$token_name} = {  map { my $key = $_; my $value;
                                    ( $key, sub { defined $value and return $value;
                                                  my $flash = session($session_hash_key) || {};
                                                  $value = delete $flash->{$key};
                                                  session $session_hash_key, $flash;
                                                  return $value;
                                              } );
                                } ( keys %{session($session_hash_key) || {} })
                           };
};

register_plugin;

1;



=pod

=head1 NAME

Dancer::Plugin::FlashMessage - Dancer plugin to display temporary messages, so called "flash messages".

=head1 VERSION

version 0.313

=head1 DESCRIPTION

This plugin helps you display temporary messages, so called "flash messages".
It provides a C<flash()> method to define the message. The plugin then takes
care of attaching the content to the session, propagating it to the templating
system, and then removing it from the session.

However, it's up to you to have a place in your views or layout where the
message will be displayed. But that's not too hard (see L<SYNOPSYS>).

Basically, the plugin gives you access to the 'flash' hash in your views. It
can be used to display flash messages.

By default, the plugin works using a decent configuration. However, you can
change the behaviour of the plugin. See L<CONFIGURATION>

=head1 NAME

Dancer::Plugin::FlashMessage - A plugin to display "flash messages" : short temporary messages

=head1 SYNOPSYS

Example with Template Toolkit

In your configuration, make sure you have session configured. Of course you can
use any session engine :

  session: "simple"

In your index.tt view or in your layout :

  <% IF flash.error %>
    <div class=error> <% flash.error %> </div>
  <% END %>

In your css :

  .error { background: #CEE5F5; padding: 0.5em;
           border: 1px solid #AACBE2; }

In your Dancer App :

  package MyWebService;

  use Dancer;
  use Dancer::Plugin::FlashMessage;

  get '/hello' => sub {
      flash error => 'Error message';
      template 'index';
  };

=head1 METHODS

=head2 flash

  # sets the flash message for the warning key
  flash warning => 'some warning message';

  # retrieves and removes the flash message for the warning key
  my $warning_message = flash 'warning';

This method can take 1 or 2 parameters. When called with two parameters, it
sets the flash message for the given key.

When called with one parameter, it returns the value of the flash message of
the given key. The message is deleted from the flash hash in the session.

In both cases, C<flash> always returns the value;

=head1 IN YOUR TEMPLATE

After having set a flash message using C<flash> in your Dancer route, you can
access the flash message from within your template. The plugin provides you
with the C<flash> hashref, that you can access in your template, for example
like this :

  <div class=error> <% flash.error %> </div>

When you use it in your template, the flash message is deleted. So next
time, C<flash.error> will not exist.

=head1 CONFIGURATION

With no configuration whatsoever, the plugin will work fine, thus contributing
to the I<keep it simple> motto of Dancer.

=head2 configuration default values

These are the default values. See below for a description of the keys

  plugins:
    FlashMessage:
      token_name: flash
      session_hash_key: _flash

=head2 configuration description

=over

=item token_name

The name of the template token that will contain the hash of flash messages.
B<Default> : C<flash>

=item session_hash_key

You probably don't need that, but this setting allows you to change the name of
the session key used to store the hash of flash messages. It may be useful in
the unlikely case where you have key name conflicts in your session. B<Default> :
C<_flash>

=back

=head1 COPYRIGHT

This software is copyright (c) 2011 by Damien "dams" Krotkine <dams@cpan.org>.

=head1 LICENCE

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=head1 AUTHORS

This module has been written by Damien "dams" Krotkine <dams@cpan.org>.

=head1 SEE ALSO

L<Dancer>

=head1 AUTHOR

Damien "dams" Krotkine

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Damien "dams" Krotkine.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Dancer plugin to display temporary messages, so called "flash messages".',
  'AUTHOR' => 'Damien "dams" Krotkine',
  'BUILD_REQUIRES' => {
    'Dancer::Test' => '0',
    'English' => '0',
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31',
    'Module::Build' => '0.3601'
  },
  'DISTNAME' => 'Dancer-Plugin-FlashMessage',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Dancer::Plugin::FlashMessage',
  'PREREQ_PM' => {
    'Carp' => '0',
    'Dancer' => '1.3001',
    'Dancer::Plugin' => '0'
  },
  'VERSION' => '0.313',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



